compiler/compiler.go

```diff
@@ -1,210 +1,136 @@
+		case e.Service != nil:
+			p.services = append(p.services, e.Service)
+		case e.Option != nil:
+			p.options = append(p.options, e.Option)
+		case e.Extend != nil:
+			p.extends = append(p.extends, e.Extend)
+		default:
+			return nil, fmt.Errorf("%s: cannot interpret Entry", e.Pos)
+		}
 	}
+	return p, nil
 }
 
-func (b *messageBuilder) addEntry(e *parser.MessageEntry) {
-	switch {
-	case e.Field != nil:
-		b.buildField(e.Field)
-	case e.Message != nil:
-		m := newMessage(e.Message, b.proto3, b.scope, b.types)
-		b.messageDesc.NestedType = append(b.messageDesc.NestedType, m)
-	case e.Enum != nil:
-	case e.Option != nil:
-	case e.Oneof != nil:
-	case e.Extend != nil:
-	case e.Reserved != nil:
-	case e.Extensions != nil:
-	default:
-		panic(fmt.Sprintf("%s: cannot interpret MessageEntry", e.Pos))
+func newProtoFromPath(file string, importPaths []string) (*Proto, error) {
+	r, err := search(file, importPaths)
+	if err != nil {
+		return nil, err
 	}
+	return newProto(file, r)
 }
 
-func (b *messageBuilder) buildField(pField *parser.Field) {
-	fieldDesc := b.createField(pField)
-	b.messageDesc.Field = append(b.messageDesc.Field, fieldDesc)
-
-	if b.proto3 && fieldDesc.Proto3Optional != nil && *fieldDesc.Proto3Optional {
-		idx := int32(len(b.messageDesc.Field)) - 1
-		fieldDesc.OneofIndex = &idx
-		name := "_" + *fieldDesc.Name
-		dcl := &pb.OneofDescriptorProto{Name: &name}
-		b.messageDesc.OneofDecl = append(b.messageDesc.OneofDecl, dcl)
+func readProtos(files, importPaths []string, done map[string]bool) ([]*Proto, error) {
+	protos := make([]*Proto, 0, len(files))
+	for _, file := range files {
+		pp, err := readProto(file, importPaths, done)
+		if err != nil {
+			return nil, err
+		}
+		protos = append(protos, pp...)
 	}
+	return protos, nil
 }
 
-var scalars = map[parser.Scalar]pb.FieldDescriptorProto_Type{
-	parser.Double:   pb.FieldDescriptorProto_TYPE_DOUBLE,
-	parser.Float:    pb.FieldDescriptorProto_TYPE_FLOAT,
-	parser.Int32:    pb.FieldDescriptorProto_TYPE_INT32,
-	parser.Int64:    pb.FieldDescriptorProto_TYPE_INT64,
-	parser.Uint32:   pb.FieldDescriptorProto_TYPE_UINT32,
-	parser.Uint64:   pb.FieldDescriptorProto_TYPE_UINT64,
-	parser.Sint32:   pb.FieldDescriptorProto_TYPE_SINT32,
-	parser.Sint64:   pb.FieldDescriptorProto_TYPE_SINT64,
-	parser.Fixed32:  pb.FieldDescriptorProto_TYPE_FIXED32,
-	parser.Fixed64:  pb.FieldDescriptorProto_TYPE_FIXED64,
-	parser.SFixed32: pb.FieldDescriptorProto_TYPE_SFIXED32,
-	parser.SFixed64: pb.FieldDescriptorProto_TYPE_SFIXED64,
-	parser.Bool:     pb.FieldDescriptorProto_TYPE_BOOL,
-	parser.String:   pb.FieldDescriptorProto_TYPE_STRING,
-	parser.Bytes:    pb.FieldDescriptorProto_TYPE_BYTES,
-}
-
-func (b *messageBuilder) createField(pf *parser.Field) *pb.FieldDescriptorProto {
-	if pf.Direct == nil || pf.Direct.Type.Map != nil {
-		panic(fmt.Sprintf("%s: non-direct not implemented", pf.Pos))
+func readProto(file string, importPaths []string, done map[string]bool) ([]*Proto, error) {
```

Comment:
I think these functions could use some comments, it's not super clear to me what the difference between them all is.

Reply:
Fair. I've tried:
```
// readProtos creates ASTs for given files and their dependencies in
// order of the original files slice. By contrast, readProto creates an
// AST for a given file and the ASTs of its dependencies, listed before
// the file AST. Multiple dependencies(imports) listed in the same
// proto file are again processed in order by readProtos(). Already
// processed files are not double listed.
//
// For example:
// * f1 imports f2 which imports f3
// * g1 imports g2, g3, f3
// readProtos([]string{f1, g1}, ...)
// results in the following order: f3, f2, f1, g2, g3, f2
```

hope that's ok.